icicle-compiler/test/Icicle/Test/Arbitrary/Program.hs

```diff
@@ -4,201 +4,565 @@
+      Map.insertWith unionByEntity (eavtInputName fact) (Map.singleton (eavtEntity fact) [fact])
+
+    unionByEntity =
+      Map.unionWith (<>)
+  in
+    foldr insertByInputName Map.empty vs
+
+-- | This is used to ensure the Core evaluation semnatics matches the C code.
+--
+evalWellTyped :: WellTypedEval -> WellTyped -> Map Entity [(OutputId, BaseValue)]
+evalWellTyped ctx wt =
+  let
+    facts =
+      wtFacts wt
+  in
+    Map.unionsWith (<>) .
+    flip fmap (wtClusters wt) $ \cluster ->
+      evalWellTypedCluster ctx cluster facts
+
+evalWellTypedCluster ::
+     WellTypedEval
+  -> WellTypedCluster
+  -> [WellTypedValue]
+  -> Map Entity [(OutputId, BaseValue)]
+evalWellTypedCluster eval cluster allFacts =
+  let
+    evalCore baseValues =
+      let
+        asStreamValue (AsAt v t) =
+          AsAt (BubbleGumFact (Flavour 0 t), v) t
+      in
+        case Core.eval (wtEvalContext eval) (fmap asStreamValue baseValues) (wtCore cluster) of
+          Left e ->
+            Savage.error . show . vsep $
+              [ "Impossible! Failed to evaluate well-typed Core with this input: "
+              , pretty e
+              , vsep (fmap (text . show) baseValues)
+              ]
+          Right xs ->
+            xs
+
+    clusterEntities =
+      List.nub . fmap eavtEntity $ allFacts
+
+    clusterInputName =
+      inputName . Sea.clusterInputId . wtCluster $ cluster
+
+    -- drop all facts for any entity that exceed the limit *for any attribute*
+
+    factsForAllInputs =
+      factsGroupedByInputName allFacts
+
+    dropFacts =
+      flip fmap factsForAllInputs $ \facts ->
+        Map.filter ((> wtEvalFactsLimit eval) . length) facts
+
+    dropEntities =
+      concatMap Map.keys . Map.elems $ dropFacts
+
+    -- reads facts for this input id that do not exceed the fact limit
+
+    factsForThisInput =
+      fromMaybe Map.empty .
+      Map.lookup clusterInputName $
+        factsForAllInputs
+
+    computeFacts =
+      fmap sortFacts .
+      Map.filter ((<= wtEvalFactsLimit eval) . length) $
+        factsForThisInput
+
+    -- NOTE (invariant)
+    -- any entity that was read has an output in psv/zebra
+    -- the output handles this logic so we need to emulate that here.
+
+    computeResults =
+      fmap (Core.value . evalCore . fmap eavtValue) computeFacts
+
+    defaultOuputsForCluster =
+      Core.value . evalCore $ []
+
+    hasOutput v =
+      case v of
+        VLeft (VError ExceptNotAnError) ->
+          True
+        VLeft _ ->
+          False
+        _ ->
+          True
+
+    hasOutputs =
+      flip Map.filterWithKey computeResults $ \k v ->
+        not (k `elem` dropEntities) &&
+        not (null (List.filter (hasOutput . snd) v))
+
+    noOutputs =
+      flip List.filter clusterEntities $ \e ->
+        not (e `elem` dropEntities) &&
+        not (e `elem` Map.keys hasOutputs)
+
+    fillers =
+      Map.fromList . fmap (, defaultOuputsForCluster) $ noOutputs
+
+  in
+    hasOutputs <> fillers
```

Comment:
❓ 
this is the bridge between Core evaluation and C evaluation.

They are the same except some logic is embedded into the input/output code (for both PSV and Zebra) that changes what facts go through C eval. This should cover all of them so that Core eval outputs the same things as C eval.